ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -51,27 +57,52 @@
   private RandomAccessFile stream;
   private FileChannel fileChannel;
   private File file = new File("client-data-stream");
+  private List<DataStreamClientImpl> clients = new ArrayList<>();
+  private List<DataStreamOutput> streams = new ArrayList<>();
 
 
   public NettyServerStreamRpc(RaftPeer server){
     this.raftServer = server;
     setupServer();
   }
 
+  public NettyServerStreamRpc(
+      RaftPeer server, List<RaftPeer> otherPeers, RaftProperties properties){
+    this.raftServer = server;
+    setupServer();
+    setupClient(otherPeers, properties);
+  }
+
   private ChannelInboundHandler getServerHandler(){
     return new ChannelInboundHandlerAdapter(){
       @Override
       public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         final DataStreamRequestByteBuf req = (DataStreamRequestByteBuf)msg;
+
+        List<CompletableFuture<DataStreamReply>> futures = new ArrayList<>();
+        // forward requests to other stream servers.
+        for (DataStreamOutput streamOutput : streams) {
+          CompletableFuture<DataStreamReply> future =
+              streamOutput.streamAsync(req.getBuf().nioBuffer());
+          futures.add(future);
+        }
+
         ByteBuffer[] bfs = req.getBuf().nioBuffers();
         for(int i = 0; i < bfs.length; i++){
           fileChannel.write(bfs[i]);
         }
-        req.getBuf().release();
-        final DataStreamReply reply = new DataStreamReplyByteBuffer(req.getStreamId(),
-                                                        req.getDataOffset(),
-                                                        ByteBuffer.wrap("OK".getBytes()));
-        ctx.writeAndFlush(reply);
+
+        try {
+          for (CompletableFuture<DataStreamReply> future : futures) {
+              future.join();
```

Comment:
At this moment, I am not sure how this server should handle exceptions when forwarding message to peers:
1. Will there any exception be thrown? I have this question because existing request handling code seems does not throw any exception already.
2. If this server fails to forward a message to a peer, what is the proper way to respond to client? Note that the client is sending a stream of message. So is failures happen in the middle means the client should retry from the failure point?

Reply:
Read should not wait for the future, otherwise it will slow down read.  We should have another thread pool to wait, handle exceptions/retries and send replies.

Reply:
I am little confused here: this is an implementation that forwards messages to peers from a server, and then wait the writes on peers finish. How is it related to read?

Reply:
my current understanding is `ctx.writeAndFlush(reply);` will complete the future on client side, so before run this code, we need to wait writes on other peers to complete. That's why I was thinking we need some sort of `future.join` before ``ctx.writeAndFlush(reply);` `

Reply:
@runzhiwang how do you think about discussion here? 

basically the requirement is:
1. forward writes to peers (which can happen in parallel).
2. The `Future` held by client says complete means no only writes on this server is completed,  but also other peers completes writes.

That's why I was thinking we might need `join` before `ctx.writeAndFlush(reply)` to wait other peers finish writes.

Reply:
@amaliujia The point of @szetszwo is there are many sub write requests of one big write request, for example, when client want to write 640KB data, it will split it into 10 sub requests, each sub request write 64KB, when server1 receive the [0, 64KB), it will forward [0, 64KB) to server2 and server3, but if we do future.join, before finish forwarding [0, 64KB), server1 can not read [64KB, 128KB), it will delay the streaming.

Reply:
Thanks @runzhiwang 

Let me think about this and make changes accordingly.

Reply:
I have to `allOf` to achieve the idea that make sure both local writes and writes to peers succeed before sending a reply.